clients/src/main/java/org/apache/kafka/common/record/MemoryRecords.java

```diff
@@ -195,11 +198,13 @@ private static FilterResult filterTo(TopicPartition partition, Iterable<MutableR
             }
 
             if (!retainedRecords.isEmpty()) {
-                if (writeOriginalBatch) {
+                // we check if the delete horizon should be set to a new value
+                // in which case, we need to reset the base timestamp and overwrite the timestamp deltas
+                if (writeOriginalBatch && !shouldSetDeleteHorizon) {
```

Comment:
We only need to write the DeleteHorizonTime if the batch contains a tombstone or a control record. This increases the chance for more optimized writeOriginalBatch case to happen.